src/engine.cc

```diff
@@ -0,0 +1,83 @@
+#include "engine.h"
+#include "gobject.h"
+
+#include <eos-knowledge-content.h>
+
+using namespace v8;
+
+struct CallbackData {
+  Persistent<Promise::Resolver> resolver;
+};
+
+static void EngineGetObjectCallback(GObject *source,
+                                    GAsyncResult *result,
+                                    gpointer user_data)
+{
+  Isolate *isolate = Isolate::GetCurrent ();
+  HandleScope scope(isolate);
+  CallbackData *data = (CallbackData *)user_data;
+
+  GError *error = NULL;
+  EkncContentObjectModel *model;
+
+  Local<Promise::Resolver> resolver = Local<Promise::Resolver>::New(isolate, data->resolver);
+  if ((model = eknc_engine_get_object_finish (EKNC_ENGINE (source), result, &error)))
+    resolver->Resolve(WrapperFromGObject (isolate, G_OBJECT (model)));
+  else
+    resolver->Reject(Exception::TypeError(String::NewFromUtf8(isolate, error->message)));
+
+  /* XXX: For some reason we have to resolve microtasks to the promise wrappers to resolve. */
+  isolate->RunMicrotasks();
+  data->resolver.Reset();
+  delete data;
+}
+
+void EngineGetObject(const FunctionCallbackInfo<Value> &args) {
+  Isolate *isolate = args.GetIsolate ();
+  GObject *engine = GObjectFromWrapper (args.This ());
+  String::Utf8Value id (args[0]->ToString ());
+  String::Utf8Value app_id (args[1]->ToString ());
+
+  Local<Promise::Resolver> resolver = Promise::Resolver::New(isolate);
+  Local<Promise> promise = resolver->GetPromise();
+  CallbackData *data = new CallbackData();
```

Comment:
Can you just allocate a `Persistent<Promise::Resolver>` on the heap directly, and get rid of `CallbackData`?